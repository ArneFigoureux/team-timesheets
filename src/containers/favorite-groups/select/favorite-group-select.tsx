import * as React from 'react';
import { observer } from 'mobx-react-lite';

import { Select } from '@rmwc/select';
import { useFavoriteGroupStore } from '../../../contexts/favorite-context/favorite-context';

export const FavoriteGroupSelect = observer(({
    value,
    onChange,
}: {
    value: string | undefined,
    onChange(value: string): void,
}) => {
    const favoriteGroupStore = useFavoriteGroupStore();

    const onGroupChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
        const value = event.currentTarget.value;
        onChange(value);
    }

    const groups = favoriteGroupStore.groups.map(
        ({ id, name }) => (
            <option
                value={id}
                key={id}>
                {name}
            </option>
        )
    );

    return (
        <Select
            value={value || ""}
            outlined
            label="Favorite group"
            onChange={onGroupChange}>
            {groups}
        </Select>
    );
});
