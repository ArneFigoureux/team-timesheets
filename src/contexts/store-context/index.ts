export * from "./store-context";
export * from "./store-provider";
export * from "./use-store";
