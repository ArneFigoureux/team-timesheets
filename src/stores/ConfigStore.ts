import { observable, computed } from 'mobx';
import { Collection, ICollection } from "firestorable";
import { IRootStore } from './RootStore';
import { IProject, ITask, IClient, ITeam } from '../../common/dist';
import { firestore } from '../firebase/myFirebase';

export interface IConfigStore {
    projects: ICollection<IProject>;
    tasks: ICollection<ITask>;
    clientsCollection: ICollection<IClient>;
    teamsCollection: ICollection<IClient>;
    clients: (IClient & {id: string})[];
    teams: (ITeam & {id: string, isSelected: boolean})[];
    taskId?: string;
    projectId?: string;
    clientId?: string;
    teamId?: string;
}

export class ConfigStore implements IConfigStore {
    //private readonly _rootStore: IRootStore;

    readonly projects: ICollection<IProject>;
    readonly tasks: ICollection<IProject>;
    readonly clientsCollection: ICollection<IClient>;
    readonly teamsCollection: ICollection<ITeam>;

    @observable.ref taskId?: string;
    @observable.ref projectId?: string;
    @observable.ref clientId?: string;
    @observable.ref teamId?: string;

    constructor(_rootStore: IRootStore, getCollection: (name: string) => firebase.firestore.CollectionReference) {
        // this._rootStore = rootStore;
        this.projects = observable(new Collection<IProject>(firestore, getCollection.bind(this, "projects"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));

        this.tasks = observable(new Collection<ITask>(firestore, getCollection.bind(this, "tasks"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));

        this.clientsCollection = observable(new Collection<ITask>(firestore, getCollection.bind(this, "clients"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));

        this.teamsCollection = observable(new Collection<ITeam>(firestore, getCollection.bind(this, "teams"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));
    }

    @computed
    public get clients() {
        return Array.from(this.clientsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id }));
    }

    @computed
    public get teams() {
        return Array.from(this.teamsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id, isSelected: doc.id === this.teamId }));
    }
}