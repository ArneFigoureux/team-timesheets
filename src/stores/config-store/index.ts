export * from "./config-store";
export * from "./use-config-store";