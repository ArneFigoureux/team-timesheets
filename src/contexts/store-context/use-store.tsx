import React from "react";
import { StoreContext } from "./store-context";

export const useStore = () => React.useContext(StoreContext);