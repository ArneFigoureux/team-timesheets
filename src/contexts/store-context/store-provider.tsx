import React from "react";
import { useFirebase } from "../firebase-context/firebase-context";
import { StoreContext } from "./store-context";
import { Store } from "../../stores/root-store";

export const StoreProvider = ({ children }: React.PropsWithChildren<{}>) => {
    const firebase = useFirebase();

    const auth = firebase.auth();
    const firestore = firebase.firestore();
    const storage = firebase.storage();

    const store = (window as any)["store"] = new Store({
        auth,
        firestore,
        storage,
    });

    return (
        <StoreContext.Provider value={store}>
            {children}
        </StoreContext.Provider>
    );
}
