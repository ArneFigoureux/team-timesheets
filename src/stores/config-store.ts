import { observable, computed } from 'mobx';
import { Collection, ICollection, RealtimeMode, FetchMode } from "firestorable";
import { IRootStore } from './root-store';
import { firestore } from '../firebase/my-firebase';
import { IProject, ITask, IClient, IClientData, ITeam, ITeamData, ITaskData, IConfig, ConfigValue } from '../../common/dist';

import * as serializer from '../../common/serialization/serializer';
import * as deserializer from '../../common/serialization/deserializer';
import { LoginProvider } from '../firebase/types';

export type Configs = {
    // authClientId: string;
    loginProviders: LoginProvider[];
}

export interface IConfigStore extends ConfigStore {};

export class ConfigStore implements IConfigStore {
    //private readonly _rootStore: IRootStore;
    readonly tasks: ICollection<IProject, ITaskData>;
    readonly clientsCollection: ICollection<IClient>;
    readonly teamsCollection: ICollection<ITeam, ITeamData>;
    readonly configsCollection: Collection<IConfig>;

    @observable.ref taskId?: string;
    @observable.ref clientId?: string;
    @observable.ref teamId?: string;

    constructor(_rootStore: IRootStore) {
        // this._rootStore = rootStore;
        const deps = { logger: console.log };

        this.tasks = new Collection<ITask, ITaskData>(
            firestore,
            "tasks",
            {
                realtimeMode: RealtimeMode.on,
                fetchMode: FetchMode.once,
                query: ref => ref.orderBy("name_insensitive"),
                serialize: serializer.convertNameWithIcon,
                deserialize: deserializer.convertNameWithIcon,
            },
            deps,
        );

        this.clientsCollection =
            new Collection<IClient, IClientData>(
                firestore,
                "clients",
                {
                    realtimeMode: RealtimeMode.on,
                    fetchMode: FetchMode.once,
                    query: ref => ref.orderBy("name_insensitive"),
                    serialize: serializer.convertNameWithIcon,
                    deserialize: deserializer.convertNameWithIcon,
                },
                deps,
            );

        this.teamsCollection = new Collection<ITeam, ITeamData>(
            firestore,
            "teams",
            {
                realtimeMode: RealtimeMode.on,
                fetchMode: FetchMode.once,
                query: ref => ref.orderBy("name_insensitive"),
                serialize: serializer.convertNameWithIcon,
                deserialize: deserializer.convertNameWithIcon,
            },
            deps,
        );

        this.configsCollection = new Collection<IConfig>(
            firestore,
            "configs",
            {
                fetchMode: FetchMode.once,
            }
        )
    }

    @computed
    public get clients() {
        return Array.from(this.clientsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id }));
    }

    @computed
    public get teams() {
        return Array.from(this.teamsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id, isSelected: doc.id === this.teamId }));
    }

    // To investigate: does getConfigValue needs mobx @computed attribute?
    public getConfigValue<T>(key: string): T;
    public getConfigValue<T>(key: string, isRequired: false): T | undefined;
    public getConfigValue<T extends ConfigValue>(key: string, isRequired = true): T | undefined {
        const doc = this.configsCollection.docs.find(c => c.data!.key === key);

        if (!doc) {
            if (isRequired) throw new Error(`Required config '${key}' is missing.`);
            else return undefined;
        }

        return doc.data!.value as unknown as T;
    }
}
