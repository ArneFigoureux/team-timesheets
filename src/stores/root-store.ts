import { IRegistrationsStore, RegistrationStore } from "./registration-store";
import { IConfigStore, ConfigStore } from "./config-store";
import { IUserStore, UserStore } from "./user-store";
import { IViewStore, ViewStore } from "./view-store";
import { RouterStore } from "mobx-router";
import { IReportStore, ReportStore } from "./report-store";
import { DashboardStore, IDashboardStore } from "./dashboard-store";
import { firestore } from "../firebase/my-firebase";
import { IProjectStore, ProjectStore } from "./project-store";

export interface IRootStore {

    readonly user: IUserStore;
    readonly view: IViewStore;
    readonly router: RouterStore;
    readonly timesheets: IRegistrationsStore;
    readonly reports: IReportStore;
    readonly config: IConfigStore;
    readonly projects: IProjectStore;
    readonly getCollection: (name: string) => firebase.firestore.CollectionReference;
}

export class Store implements IRootStore {
    public readonly timesheets: IRegistrationsStore;
    public readonly view: IViewStore;
    public readonly user: IUserStore;
    public readonly config: IConfigStore;
    public readonly router = new RouterStore();
    public readonly reports: IReportStore;
    public readonly dashboard: IDashboardStore;
    public readonly projects: IProjectStore;

    public readonly getCollection: (name: string) => firebase.firestore.CollectionReference;

    constructor(getCollection: (name: string) => firebase.firestore.CollectionReference) {
        this.getCollection = getCollection;

        this.user = new UserStore(this);
        this.view = new ViewStore(this);
        this.config = new ConfigStore(this, getCollection);
        this.timesheets = new RegistrationStore(this);
        this.reports = new ReportStore(this);
        this.dashboard = new DashboardStore(this);
        this.projects = new ProjectStore(this);

    }
};

const store = (window as any)["store"] = new Store(name => firestore.collection(name));

export default store;

