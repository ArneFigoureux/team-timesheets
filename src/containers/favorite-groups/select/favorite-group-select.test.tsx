import React from "react";
import fs from "fs";
import path from "path";
import type firebase from "firebase";
import { render, waitFor } from "@testing-library/react";
import { loadFirestoreRules, initializeTestApp, clearFirestoreData } from "@firebase/rules-unit-testing";
import { Store } from "../../../stores/root-store";
import { useStore } from "../../../contexts/store-context";
import userEvent from "@testing-library/user-event";
import { FavoriteGroupSelect } from "./favorite-group-select";

jest.mock("../../../contexts/store-context");

const projectId = "project-select-test";
const app = initializeTestApp({
    projectId,
});

beforeAll(async () => {
    await loadFirestoreRules({
        projectId,
        rules: fs.readFileSync(path.resolve(__dirname, "../../../../firestore.rules.test"), "utf8"),
    });
});

const userId = "user-1";
let store: Store;
beforeEach(async () => {
    store = new Store({
        firestore: app.firestore(),
    });

    await Promise.all([
        store.user.usersCollection.addAsync(
            {
                name: "user 1",
                team: "team-1",
                roles: {
                    user: true,
                },
                uid: userId,
                tasks: new Map(),
                recentProjects: [],
                divisionId: "",
            },
            userId,
        ),
    ]);

    store.auth.setUser({
        uid: "user-1",
    } as firebase.User);

    (useStore as jest.Mock<ReturnType<typeof useStore>>).mockReturnValue(store);
});

afterEach(async () => {
    store.dispose();
    await clearFirestoreData({ projectId });
});

afterAll(() => app.delete());

describe("FavoriteGroupSelect", () => {
    it("should render without favorite groups", () => {
        const {
            asFragment,
        } = render(
            <FavoriteGroupSelect
                onChange={jest.fn()}
                value={undefined}
            />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    describe("when there are favorite groups", () => {
        let groupIds: string[];
        beforeEach(async () => {
            groupIds = await store.favorites.addDocuments([
                {
                    name: "favorite group 1",
                    userId,
                },
                {
                    name: "favorite group 2",
                    userId,
                },
                {
                    name: "favorite group 3",
                    userId,
                },
            ]);
        });

        it("should show an option for each favorite group", async () => {
            const {
                getByText,
                unmount,
            } = render(
                <FavoriteGroupSelect
                    onChange={jest.fn()}
                    value={undefined}
                />
            );

            await waitFor(() => getByText("favorite group 1"));
            await waitFor(() => getByText("favorite group 2"));
            await waitFor(() => getByText("favorite group 3"));

            unmount();
        });

        it("should call onChange when favorite group is selected", async () => {
            const onChange = jest.fn();
            const {
                container,
                getByText,
                unmount,
            } = render(
                <FavoriteGroupSelect
                    value={undefined}
                    onChange={onChange}
                />
            );

            await waitFor(() => getByText("favorite group 3"));

            userEvent.selectOptions(container.querySelector("select")!, [groupIds[2]]);

            expect(onChange).toBeCalledWith(groupIds[2]);

            unmount();
        });

        it("should select the given favorite group", async () => {
            const {
                container,
                unmount,
            } = render(
                <FavoriteGroupSelect
                    value={groupIds[2]}
                    onChange={jest.fn()}
                />
            );

            await waitFor(
                () => expect(
                    container.querySelector("select")?.value
                ).toBe(groupIds[2]),
            );

            unmount();
        });
    });
});
