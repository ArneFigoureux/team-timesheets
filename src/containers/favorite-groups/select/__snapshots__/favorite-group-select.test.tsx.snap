// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteGroupSelect should render without favorite groups 1`] = `
<DocumentFragment>
  <div
    class="mdc-select mdc-select--outlined"
    role="listbox"
  >
    <div
      class="mdc-select__anchor"
    >
      <i
        class="mdc-select__dropdown-icon"
      />
      <div
        aria-disabled="false"
        aria-haspopup="listbox"
        class="mdc-select__selected-text"
        role="button"
        tabindex="-1"
      >
        Â 
      </div>
      <div
        class="mdc-notched-outline mdc-notched-outline--upgraded"
      >
        <div
          class="mdc-notched-outline__leading"
        />
        <div
          class="mdc-notched-outline__notch"
        >
          <span
            class="mdc-floating-label"
            for="select-Favorite group"
            style="transition-duration: 0s;"
          >
            Favorite group
          </span>
        </div>
        <div
          class="mdc-notched-outline__trailing"
        />
      </div>
      <select
        class="rmwc-select__native-control "
        tabindex="0"
      >
        <option
          disabled=""
          value=""
        />
      </select>
    </div>
  </div>
</DocumentFragment>
`;
